astropy/nddata/mixins/ndarithmetic.py

```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
 
-    def multiply(self, operand, propagate_uncertainties=True):
-        if propagate_uncertainties:
-            propagate_uncertainties = "propagate_multiply"
-        else:
-            propagate_uncertainties = None
-        return self._arithmetic(
-            operand, propagate_uncertainties, "multiplication", np.multiply)
+        Parameters
+        ----------
+        operation: `str`
+            see :meth:`NDArithmeticMixin.arithmetic` parameter description.
+
+        operand: `NDData`-like instance
+            The second operand wrapped in an instance of the same class as
+            self.
+
+        meta_kwds_operate: ``None`` or `list`
+            see :meth:`NDArithmeticMixin.arithmetic` parameter description.
 
-    if isinstance(_arithmetic.__doc__, string_types):
-        multiply.__doc__ = _arithmetic.__doc__.format(name="Multiply",
-                                                      operator="*")
+        meta_kwds_set: ``None`` or `dict`
+            see :meth:`NDArithmeticMixin.arithmetic` parameter description.
 
-    def divide(self, operand, propagate_uncertainties=True):
-        if propagate_uncertainties:
-            propagate_uncertainties = "propagate_divide"
+        Returns
+        -------
+        result_meta: `dict`-like
+            The class of the result depends on which meta is set. If only the
+            second operand has a not empty meta property the return has the
+            same class as the second operands meta. In all other cases the
+            result has the same class as the first operands meta.
+        """
+
+        # Again this is only partly arithmetics. But it should be best to
+        # keep everything relating meta in one operation.
+
+        # Starting by creating the basic meta for the result
+        if len(self.meta) == 0 and len(operand.meta) > 0:
+            # The only case in which we would like the meta from the second
+            # operand is if the first operand has no meta and the second one
+            # contains at least one meta element.
+            # There is no use-case where merging the meta of both operands
+            # makes sense. In case one might really set the results meta
+            # information to match the one of the second operand is to put a
+            # reference or copy inside the parameter that sets the keywords
+            # in here.
+            result_meta = deepcopy(operand.meta)
         else:
-            propagate_uncertainties = None
-        return self._arithmetic(
-            operand, propagate_uncertainties, "division", np.divide)
+            result_meta = deepcopy(self.meta)
+
+        # Now get to work, first we check if there are any keywords which
+        # should be arithmetically operated upon.
+        if meta_kwds_operate is not None and len(meta_kwds_operate) > 0:
+
+            # Get the operator function for these operation (use numpy ufuncs)
+            if operation == 'addition':
```

Comment:
This definitely makes the case using functions as the argument instead of strings!